{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       ":get_entryid"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "class Gene\n",
    "\n",
    "  attr_accessor :geneid\n",
    "  attr_accessor :kegg\n",
    "  attr_accessor :go\n",
    "  @@number_of_genes = 0\n",
    "  @@genes = Hash.new\n",
    "  \n",
    "\n",
    "  \n",
    "  def initialize (params = {}) \n",
    "      @geneid = params.fetch(:geneid, 'nil')\n",
    "      @kegg = params.fetch(:kegg, 'nil')\n",
    "      @go = params.fetch(:go, 'nil')\n",
    "    \n",
    "      @@number_of_objects =+ 1\n",
    "      @@genes[geneid] = self\n",
    "    \n",
    "  end\n",
    "  \n",
    "  \n",
    "  \n",
    "  def Gene.create_object(thisfile)\n",
    "    File.readlines(thisfile).each do |line|\n",
    "    geneid = self.get_geneid(line)\n",
    "    entry_id = self.get_entryid(line)\n",
    "    accession = self.get_accesions(line)\n",
    "      \n",
    "    Gene.new(\n",
    "        :geneid => geneid,\n",
    "        :accession => accession,\n",
    "        :entry_id => entry_id)\n",
    "    end\n",
    "  end\n",
    "  \n",
    "  def self.get_geneid(agi_code)\n",
    "    \n",
    "    match_id = Regexp.new(/A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/)\n",
    "      if match_id.match(agi_code)\n",
    "        geneid = agi_code\n",
    "      else \n",
    "        puts \"wrong AGI code\"\n",
    "      end\n",
    "    \n",
    "    return geneid \n",
    "  end\n",
    "    \n",
    "  def self.get_information_kegg(agi_code)\n",
    "    \n",
    "    address = \"http://togows.org/entry/ebi-uniprot/#{agi_code}/accessions.json\"\n",
    "    response = RestClient::Request.execute(\n",
    "    method: :get,\n",
    "    url: address)\n",
    "    \n",
    "    accessions = JSON.parse(response.body)\n",
    "    return accessions[0]\n",
    "  end\n",
    "  \n",
    "http://www.ebi.ac.uk/Tools/webservices/psicquic/intact/webservices/current/search/interactor/P51587?format=tab25\n",
    "  \n",
    "\n",
    "  def self.get_entryid(agi_code)\n",
    "    \n",
    "    address = \"http://togows.org/entry/ebi-uniprot/#{agi_code}/entry_id\"\n",
    "    response = RestClient::Request.execute(\n",
    "    method: :get,\n",
    "    url: address)\n",
    "    \n",
    "    entry_id = JSON.parse(response.body)\n",
    "\n",
    "    return entry_id\n",
    "  end\n",
    "    \n",
    "end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_object_2(agi_code)\n",
    "#     if Gene.get_all_objects[\"#{agi_code}\"].gene_depth == 1\n",
    "      Gene.get_all_objects[\"#{agi_code}\"].int_array.each do |gene|\n",
    "        go = self.get_go(gene)\n",
    "        puts go\n",
    "        kegg = self.get_kegg(gene)\n",
    "        puts kegg\n",
    "        intact = self.get_intact_array(gene)\n",
    "        puts intact\n",
    "        gene_depth = Gene.get_all_objects[\"#{agi_code}\"].gene_depth + 1\n",
    "        \n",
    "        Gene.new(\n",
    "        :geneid => geneid,\n",
    "        :go => go,\n",
    "        :kegg => kegg,\n",
    "        :int_array => intact,\n",
    "        :gene_depth => gene_depth)\n",
    "      end\n",
    "        \n",
    "    end\n",
    "        \n",
    "        \n",
    "  end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "ename": "JSON::ParserError",
     "evalue": "767: unexpected token at '{\"EMBL\"=>[[\"AF143691\", \"AAD37362.1\", \"-\", \"mRNA\"], [\"AF134126\", \"AAD28773.1\", \"-\", \"mRNA\"], [\"AB010695\", \"BAB10750.1\", \"-\", \"Genomic_DNA\"], [\"CP002688\", \"AED96477.1\", \"-\", \"Genomic_DNA\"], [\"AF361858\", \"AAK32870.1\", \"-\", \"mRNA\"], [\"AF372917\", \"AAK49633.1\", \"-\", \"mRNA\"], [\"AY057735\", \"AAL15365.1\", \"-\", \"mRNA\"], [\"AK318653\", \"BAH56768.1\", \"-\", \"mRNA\"], [\"Z18479\", \"CAA79201.1\", \"-\", \"mRNA\"], [\"GL348720\", \"EFH40582.1\", \"-\", \"Genomic_DNA\"], [\"LUHQ01000005\", \"OAO95972.1\", \"-\", \"Genomic_DNA\"], [\"LUHQ01000005\", \"OAO89788.1\", \"-\", \"Genomic_DNA\"]], \"PIR\"=>[[\"T52318\", \"T52318\"]], \"RefSeq\"=>[[\"NP_200238.1\", \"NM_124807.4. [Q9S7M0-1\"], [\"XP_002864323.1\", \"XM_002864277.1\"], [\"NP_200238.1\", \"NM_124807.4\"], [\"NP_568821.1\", \"NM_124896.3\"]], \"SMR\"=>[[\"Q9S7M0\", \"-\"], [\"A0A178USW4\", \"-\"], [\"A0A178U792\", \"-\"]], \"STRING\"=>[[\"3702.AT5G54270.1\", \"-\"], [\"59689.fgenesh2_kg.8__1328__AT5G54270.1\", \"-\"]], \"iPTMnet\"=>[[\"Q9S7M0\", \"-\"]], \"PaxDb\"=>[[\"Q9S7M0\", \"-\"]], \"PRIDE\"=>[[\"Q9S7M0\", \"-\"], [\"D7MU76\", \"-\"]], \"EnsemblPlants\"=>[[\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270. [Q9S7M0-1\"], [\"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\"], [\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270\"], [\"AT5G55140.1\", \"AT5G55140.1\", \"AT5G55140\"]], \"GeneID\"=>[[\"835515\", \"-\"], [\"9300399\", \"-\"], [\"835515\", \"-\"], [\"835607\", \"-\"]], \"Gramene\"=>[[\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270. [Q9S7M0-1\"], [\"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\"], [\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270\"], [\"AT5G55140.1\", \"AT5G55140.1\", \"AT5G55140\"]], \"KEGG\"=>[[\"ath:AT5G54270\", \"-\"], [\"aly:ARALYDRAFT_495512\", \"-\"], [\"ath:AT5G54270\", \"-\"], [\"ath:AT5G55140\", \"-\"]], \"Araport\"=>[[\"AT5G54270\", \"-\"]], \"TAIR\"=>[[\"locus:2162540\", \"AT5G54270\"]], \"eggNOG\"=>[[\"ENOG410IKZW\", \"Eukaryota\"], [\"ENOG410YAI6\", \"LUCA\"]], \"HOGENOM\"=>[[\"HOG000238032\", \"-\"]], \"InParanoid\"=>[[\"Q9S7M0\", \"-\"]], \"KO\"=>[[\"K08914\", \"-\"], [\"K08914\", \"-\"], [\"K08914\", \"-\"], [\"K02907\", \"-\"]], \"OMA\"=>[[\"RITMSNE\", \"-\"], [\"RITMSNE\", \"-\"], [\"MYNARKE\", \"-\"]], \"OrthoDB\"=>[[\"986689at2759\", \"-\"], [\"986689at2759\", \"-\"], [\"986689at2759\", \"-\"], [\"1523595at2759\", \"-\"]], \"PhylomeDB\"=>[[\"Q9S7M0\", \"-\"], [\"A0A178USW4\", \"-\"]], \"PRO\"=>[[\"PR:Q9S7M0\", \"-\"]], \"Proteomes\"=>[[\"UP000006548\", \"Chromosome 5\"], [\"UP000008694\", \"Unassembled WGS sequence\"], [\"UP000078284\", \"Chromosome 5\"], [\"UP000078284\", \"Chromosome 5\"]], \"ExpressionAtlas\"=>[[\"Q9S7M0\", \"baseline and differential\"], [\"A0A178USW4\", \"baseline and differential\"], [\"A0A178U792\", \"baseline and differential\"]], \"GO\"=>[[\"GO:0009507\", \"C:chloroplast\", \"IDA:TAIR\"], [\"GO:0009941\", \"C:chloroplast envelope\", \"IBA:GO_Central\"], [\"GO:0009534\", \"C:chloroplast thylakoid\", \"IDA:TAIR\"], [\"GO:0009535\", \"C:chloroplast thylakoid membrane\", \"IDA:TAIR\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0009536\", \"C:plastid\", \"HDA:TAIR\"], [\"GO:0010287\", \"C:plastoglobule\", \"IBA:GO_Central\"], [\"GO:0009517\", \"C:PSII associated light-harvesting complex II\", \"IDA:TAIR\"], [\"GO:0009579\", \"C:thylakoid\", \"IDA:TAIR\"], [\"GO:0042651\", \"C:thylakoid membrane\", \"IDA:UniProtKB\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0046872\", \"F:metal ion binding\", \"IEA:UniProtKB-KW\"], [\"GO:0019904\", \"F:protein domain specific binding\", \"IPI:CAFA\"], [\"GO:0015979\", \"P:photosynthesis\", \"IMP:TAIR\"], [\"GO:0009768\", \"P:photosynthesis, light harvesting in photosystem I\", \"IBA:GO_Central\"], [\"GO:0009769\", \"P:photosynthesis, light harvesting in photosystem II\", \"IMP:UniProtKB\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0010119\", \"P:regulation of stomatal movement\", \"IMP:UniProtKB\"], [\"GO:0009737\", \"P:response to abscisic acid\", \"IMP:UniProtKB\"], [\"GO:0009635\", \"P:response to herbicide\", \"IEP:UniProtKB\"], [\"GO:0009644\", \"P:response to high light intensity\", \"IEP:UniProtKB\"], [\"GO:0009416\", \"P:response to light stimulus\", \"IBA:GO_Central\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0009517\", \"C:PSII associated light-harvesting complex II\", \"IEA:EnsemblPlants\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0019904\", \"F:protein domain specific binding\", \"IEA:EnsemblPlants\"], [\"GO:0009769\", \"P:photosynthesis, light harvesting in photosystem II\", \"IEA:EnsemblPlants\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0010119\", \"P:regulation of stomatal movement\", \"IEA:EnsemblPlants\"], [\"GO:0009737\", \"P:response to abscisic acid\", \"IEA:EnsemblPlants\"], [\"GO:0009635\", \"P:response to herbicide\", \"IEA:EnsemblPlants\"], [\"GO:0009644\", \"P:response to high light intensity\", \"IEA:EnsemblPlants\"], [\"GO:0009535\", \"C:chloroplast thylakoid membrane\", \"IEA:UniProtKB-SubCell\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0009765\", \"P:photosynthesis, light harvesting\", \"IEA:InterPro\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0015934\", \"C:large ribosomal subunit\", \"IEA:InterPro\"], [\"GO:0003735\", \"F:structural constituent of ribosome\", \"IEA:InterPro\"], [\"GO:0006412\", \"P:translation\", \"IEA:InterPro\"]], \"Gene3D\"=>[[\"1.10.3460.10\", \"-\", \"1\"], [\"1.10.3460.10\", \"-\", \"1\"], [\"1.10.3460.10\", \"-\", \"1\"], [\"3.30.1390.20\", \"-\", \"1\"]], \"InterPro\"=>[[\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR036919\", \"L30_ferredoxin-like_sf\"], [\"IPR005996\", \"Ribosomal_L30_bac-type\"], [\"IPR016082\", \"Ribosomal_L30_ferredoxin-like\"]], \"PANTHER\"=>[[\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR15892\", \"PTHR15892\", \"1\"]], \"Pfam\"=>[[\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00327\", \"Ribosomal_L30\", \"1\"]], \"CDD\"=>[[\"cd01658\", \"Ribosomal_L30\", \"1\"]], \"HAMAP\"=>[[\"MF_01371_B\", \"Ribosomal_L30_B\", \"1\"]], \"SUPFAM\"=>[[\"SSF55129\", \"SSF55129\", \"1\"]], \"TIGRFAMs\"=>[[\"TIGR01308\", \"rpmD_bact\", \"1\"]]}\n'",
     "output_type": "error",
     "traceback": [
      "\u001b[31mJSON::ParserError\u001b[0m: 767: unexpected token at '{\"EMBL\"=>[[\"AF143691\", \"AAD37362.1\", \"-\", \"mRNA\"], [\"AF134126\", \"AAD28773.1\", \"-\", \"mRNA\"], [\"AB010695\", \"BAB10750.1\", \"-\", \"Genomic_DNA\"], [\"CP002688\", \"AED96477.1\", \"-\", \"Genomic_DNA\"], [\"AF361858\", \"AAK32870.1\", \"-\", \"mRNA\"], [\"AF372917\", \"AAK49633.1\", \"-\", \"mRNA\"], [\"AY057735\", \"AAL15365.1\", \"-\", \"mRNA\"], [\"AK318653\", \"BAH56768.1\", \"-\", \"mRNA\"], [\"Z18479\", \"CAA79201.1\", \"-\", \"mRNA\"], [\"GL348720\", \"EFH40582.1\", \"-\", \"Genomic_DNA\"], [\"LUHQ01000005\", \"OAO95972.1\", \"-\", \"Genomic_DNA\"], [\"LUHQ01000005\", \"OAO89788.1\", \"-\", \"Genomic_DNA\"]], \"PIR\"=>[[\"T52318\", \"T52318\"]], \"RefSeq\"=>[[\"NP_200238.1\", \"NM_124807.4. [Q9S7M0-1\"], [\"XP_002864323.1\", \"XM_002864277.1\"], [\"NP_200238.1\", \"NM_124807.4\"], [\"NP_568821.1\", \"NM_124896.3\"]], \"SMR\"=>[[\"Q9S7M0\", \"-\"], [\"A0A178USW4\", \"-\"], [\"A0A178U792\", \"-\"]], \"STRING\"=>[[\"3702.AT5G54270.1\", \"-\"], [\"59689.fgenesh2_kg.8__1328__AT5G54270.1\", \"-\"]], \"iPTMnet\"=>[[\"Q9S7M0\", \"-\"]], \"PaxDb\"=>[[\"Q9S7M0\", \"-\"]], \"PRIDE\"=>[[\"Q9S7M0\", \"-\"], [\"D7MU76\", \"-\"]], \"EnsemblPlants\"=>[[\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270. [Q9S7M0-1\"], [\"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\"], [\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270\"], [\"AT5G55140.1\", \"AT5G55140.1\", \"AT5G55140\"]], \"GeneID\"=>[[\"835515\", \"-\"], [\"9300399\", \"-\"], [\"835515\", \"-\"], [\"835607\", \"-\"]], \"Gramene\"=>[[\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270. [Q9S7M0-1\"], [\"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\", \"fgenesh2_kg.8__1328__AT5G54270.1\"], [\"AT5G54270.1\", \"AT5G54270.1\", \"AT5G54270\"], [\"AT5G55140.1\", \"AT5G55140.1\", \"AT5G55140\"]], \"KEGG\"=>[[\"ath:AT5G54270\", \"-\"], [\"aly:ARALYDRAFT_495512\", \"-\"], [\"ath:AT5G54270\", \"-\"], [\"ath:AT5G55140\", \"-\"]], \"Araport\"=>[[\"AT5G54270\", \"-\"]], \"TAIR\"=>[[\"locus:2162540\", \"AT5G54270\"]], \"eggNOG\"=>[[\"ENOG410IKZW\", \"Eukaryota\"], [\"ENOG410YAI6\", \"LUCA\"]], \"HOGENOM\"=>[[\"HOG000238032\", \"-\"]], \"InParanoid\"=>[[\"Q9S7M0\", \"-\"]], \"KO\"=>[[\"K08914\", \"-\"], [\"K08914\", \"-\"], [\"K08914\", \"-\"], [\"K02907\", \"-\"]], \"OMA\"=>[[\"RITMSNE\", \"-\"], [\"RITMSNE\", \"-\"], [\"MYNARKE\", \"-\"]], \"OrthoDB\"=>[[\"986689at2759\", \"-\"], [\"986689at2759\", \"-\"], [\"986689at2759\", \"-\"], [\"1523595at2759\", \"-\"]], \"PhylomeDB\"=>[[\"Q9S7M0\", \"-\"], [\"A0A178USW4\", \"-\"]], \"PRO\"=>[[\"PR:Q9S7M0\", \"-\"]], \"Proteomes\"=>[[\"UP000006548\", \"Chromosome 5\"], [\"UP000008694\", \"Unassembled WGS sequence\"], [\"UP000078284\", \"Chromosome 5\"], [\"UP000078284\", \"Chromosome 5\"]], \"ExpressionAtlas\"=>[[\"Q9S7M0\", \"baseline and differential\"], [\"A0A178USW4\", \"baseline and differential\"], [\"A0A178U792\", \"baseline and differential\"]], \"GO\"=>[[\"GO:0009507\", \"C:chloroplast\", \"IDA:TAIR\"], [\"GO:0009941\", \"C:chloroplast envelope\", \"IBA:GO_Central\"], [\"GO:0009534\", \"C:chloroplast thylakoid\", \"IDA:TAIR\"], [\"GO:0009535\", \"C:chloroplast thylakoid membrane\", \"IDA:TAIR\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0009536\", \"C:plastid\", \"HDA:TAIR\"], [\"GO:0010287\", \"C:plastoglobule\", \"IBA:GO_Central\"], [\"GO:0009517\", \"C:PSII associated light-harvesting complex II\", \"IDA:TAIR\"], [\"GO:0009579\", \"C:thylakoid\", \"IDA:TAIR\"], [\"GO:0042651\", \"C:thylakoid membrane\", \"IDA:UniProtKB\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0046872\", \"F:metal ion binding\", \"IEA:UniProtKB-KW\"], [\"GO:0019904\", \"F:protein domain specific binding\", \"IPI:CAFA\"], [\"GO:0015979\", \"P:photosynthesis\", \"IMP:TAIR\"], [\"GO:0009768\", \"P:photosynthesis, light harvesting in photosystem I\", \"IBA:GO_Central\"], [\"GO:0009769\", \"P:photosynthesis, light harvesting in photosystem II\", \"IMP:UniProtKB\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0010119\", \"P:regulation of stomatal movement\", \"IMP:UniProtKB\"], [\"GO:0009737\", \"P:response to abscisic acid\", \"IMP:UniProtKB\"], [\"GO:0009635\", \"P:response to herbicide\", \"IEP:UniProtKB\"], [\"GO:0009644\", \"P:response to high light intensity\", \"IEP:UniProtKB\"], [\"GO:0009416\", \"P:response to light stimulus\", \"IBA:GO_Central\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0009517\", \"C:PSII associated light-harvesting complex II\", \"IEA:EnsemblPlants\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0019904\", \"F:protein domain specific binding\", \"IEA:EnsemblPlants\"], [\"GO:0009769\", \"P:photosynthesis, light harvesting in photosystem II\", \"IEA:EnsemblPlants\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0010119\", \"P:regulation of stomatal movement\", \"IEA:EnsemblPlants\"], [\"GO:0009737\", \"P:response to abscisic acid\", \"IEA:EnsemblPlants\"], [\"GO:0009635\", \"P:response to herbicide\", \"IEA:EnsemblPlants\"], [\"GO:0009644\", \"P:response to high light intensity\", \"IEA:EnsemblPlants\"], [\"GO:0009535\", \"C:chloroplast thylakoid membrane\", \"IEA:UniProtKB-SubCell\"], [\"GO:0016021\", \"C:integral component of membrane\", \"IEA:UniProtKB-KW\"], [\"GO:0009522\", \"C:photosystem I\", \"IEA:UniProtKB-KW\"], [\"GO:0009523\", \"C:photosystem II\", \"IEA:UniProtKB-KW\"], [\"GO:0016168\", \"F:chlorophyll binding\", \"IEA:UniProtKB-KW\"], [\"GO:0009765\", \"P:photosynthesis, light harvesting\", \"IEA:InterPro\"], [\"GO:0018298\", \"P:protein-chromophore linkage\", \"IEA:UniProtKB-KW\"], [\"GO:0015934\", \"C:large ribosomal subunit\", \"IEA:InterPro\"], [\"GO:0003735\", \"F:structural constituent of ribosome\", \"IEA:InterPro\"], [\"GO:0006412\", \"P:translation\", \"IEA:InterPro\"]], \"Gene3D\"=>[[\"1.10.3460.10\", \"-\", \"1\"], [\"1.10.3460.10\", \"-\", \"1\"], [\"1.10.3460.10\", \"-\", \"1\"], [\"3.30.1390.20\", \"-\", \"1\"]], \"InterPro\"=>[[\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR001344\", \"Chloro_AB-bd_pln\"], [\"IPR022796\", \"Chloroa_b-bind\"], [\"IPR023329\", \"Chlorophyll_a/b-bd_dom_sf\"], [\"IPR036919\", \"L30_ferredoxin-like_sf\"], [\"IPR005996\", \"Ribosomal_L30_bac-type\"], [\"IPR016082\", \"Ribosomal_L30_ferredoxin-like\"]], \"PANTHER\"=>[[\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR21649\", \"PTHR21649\", \"1\"], [\"PTHR15892\", \"PTHR15892\", \"1\"]], \"Pfam\"=>[[\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00504\", \"Chloroa_b-bind\", \"1\"], [\"PF00327\", \"Ribosomal_L30\", \"1\"]], \"CDD\"=>[[\"cd01658\", \"Ribosomal_L30\", \"1\"]], \"HAMAP\"=>[[\"MF_01371_B\", \"Ribosomal_L30_B\", \"1\"]], \"SUPFAM\"=>[[\"SSF55129\", \"SSF55129\", \"1\"]], \"TIGRFAMs\"=>[[\"TIGR01308\", \"rpmD_bact\", \"1\"]]}\n'",
      "\u001b[37m/home/osboxes/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/json/common.rb:156:in `parse'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/rubies/ruby-2.6.3/lib/ruby/2.6.0/json/common.rb:156:in `parse'\u001b[0m",
      "\u001b[37m(pry):1307:in `<main>'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:387:in `eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:387:in `evaluate_ruby'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:355:in `handle_line'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:274:in `block (2 levels) in eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:273:in `catch'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:273:in `block in eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:272:in `catch'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/pry-0.12.2/lib/pry/pry_instance.rb:272:in `eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/backend.rb:66:in `eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/backend.rb:12:in `eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/kernel.rb:87:in `execute_request'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/kernel.rb:47:in `dispatch'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/kernel.rb:37:in `run'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/command.rb:70:in `run_kernel'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/lib/iruby/command.rb:34:in `run'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/gems/iruby-0.3/bin/iruby:5:in `<top (required)>'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/bin/iruby:23:in `load'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/bin/iruby:23:in `<main>'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `eval'\u001b[0m",
      "\u001b[37m/home/osboxes/.rvm/gems/ruby-2.6.3/bin/ruby_executable_hooks:24:in `<main>'\u001b[0m"
     ]
    }
   ],
   "source": [
    "address= \"http://togows.org/entry/ebi-uniprot/AT5g54270/dr\"\n",
    "response = RestClient::Request.execute(\n",
    "    method: :get,\n",
    "    url: address)\n",
    "    \n",
    "entry_id = JSON.parse(response.body)\n",
    "entry_id[0][\"kegg\"]\n",
    "\"http://togows.org/entry/kegg-pathway/ath:#{agi_code}.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\"GO:0015979\"=>\"IMP:TAIR\", \"GO:0009768\"=>\"IBA:GO_Central\", \"GO:0009769\"=>\"IEA:EnsemblPlants\", \"GO:0018298\"=>\"IEA:UniProtKB-KW\", \"GO:0010119\"=>\"IEA:EnsemblPlants\", \"GO:0009737\"=>\"IEA:EnsemblPlants\", \"GO:0009635\"=>\"IEA:EnsemblPlants\", \"GO:0009644\"=>\"IEA:EnsemblPlants\", \"GO:0009416\"=>\"IBA:GO_Central\", \"GO:0009765\"=>\"IEA:InterPro\", \"GO:0006412\"=>\"IEA:InterPro\"}\n"
     ]
    }
   ],
   "source": [
    "require 'json'\n",
    "def fetch(url, headers = {accept: \"*/*\"}, user = \"\", pass=\"\") #lo mas imp es la url\n",
    "  response = RestClient::Request.execute({\n",
    "    method: :get,\n",
    "    url: url.to_s,\n",
    "    user: user,\n",
    "    password: pass,\n",
    "    headers: headers})\n",
    "  return response\n",
    "  #si algo va mal:rescue\n",
    "  rescue RestClient::ExceptionWithResponse => e\n",
    "    $stderr.puts e.response\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue RestClient::Exception => e\n",
    "    $stderr.puts e.response\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "  rescue Exception => e\n",
    "    $stderr.puts e\n",
    "    response = false\n",
    "    return response  # now we are returning 'False', and we will check that with an \\\"if\\\" statement in our main code\n",
    "end \n",
    "  \n",
    "    res = fetch(\"http://togows.org/entry/ebi-uniprot/AT5g54270/dr.json\");\n",
    "#puts res.class\n",
    "#puts res.public_methods\n",
    "  res\n",
    "if res  # res is either the response object (Class Net::HTTPSuccess), or False, so you can test it with 'if'\n",
    "  body = JSON.parse(res.body) # get the \"body\" of the response\n",
    "  all_go = body[0][\"GO\"]\n",
    "#   puts all_go\n",
    "  go_p = Hash.new\n",
    "  all_go.each do |i|\n",
    "    if i[1] =~ /P:/\n",
    "      go_p[i[0]] = i[2]\n",
    "    end\n",
    "  end\n",
    "puts go_p\n",
    "end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "prot1  AT4G09010\n",
      "prot2  AT3G53900\n",
      "prot2  AT4G09010\n",
      "no existe\n",
      "prot1  AT4G09010\n",
      "no existe\n",
      "prot2  AT4G09010\n",
      "no existe\n",
      "[\"AT3G53900\"]\n"
     ]
    }
   ],
   "source": [
    "interaction=Array.new\n",
    "    res_intact = fetch(\"http://www.ebi.ac.uk/Tools/webservices/psicquic/intact/webservices/current/search/interactor/AT4g09010?format=tab25\");\n",
    "    agi_code = \"AT4g09010\"\n",
    "    agi_code = agi_code.upcase\n",
    "      if res_intact  # res is either the response object (Class Net::HTTPSuccess), or False, so you can test it with 'if'\n",
    "        body = res_intact.body  # get the \"body\" of the response\n",
    "        lines = body.split(\"\\n\")\n",
    "        lines.each do |int|\n",
    "          #puts \"int: #{int}\"\n",
    "          #puts puts puts\n",
    "          column = int.split(\"\\t\")\n",
    "          #puts \"column: #{column}\"\n",
    "          #puts puts puts\n",
    "          match_id = Regexp.new(/A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/)\n",
    "#           column.each_with_index do |content, index|\n",
    "#             puts \"column #{index} is #{content}\"\n",
    "#           puts \"match4: #{match_id.match(column[4])}\"\n",
    "#           puts \"match5: #{match_id.match(column[5])}\"\n",
    "#           puts match_id.match(column[4]).to_s\n",
    "#            if match_id.match(column[4])\n",
    "          if match_id.match(column[4])\n",
    "             idprot1 = match_id.match(column[4]).to_s.upcase\n",
    "#              idprot1 = idprot1.upcase\n",
    "            puts \"prot1  #{idprot1}\"\n",
    "          end\n",
    "#             end\n",
    "#           if match_id.match(column[5])\n",
    "          if match_id.match(column[5])\n",
    "            idprot2 = match_id.match(column[5]).to_s.upcase\n",
    "\n",
    "          puts \"prot2  #{idprot2}\"\n",
    "          end\n",
    "#           end\n",
    "          if idprot1 && idprot2 \n",
    "            if column[0]==column[1]\n",
    "            puts \"self-interaction\"\n",
    "            interaction << agi_code\n",
    "            elsif idprot1 == agi_code\n",
    "            interaction << idprot2\n",
    "            elsif idprot2 == agi_code \n",
    "            interaction << idprot1\n",
    "            end\n",
    "          else \n",
    "            puts \"no existe\"\n",
    "          end\n",
    "        end\n",
    "      end\n",
    "puts interaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\"AT3G53900\", nil, nil, nil]\n"
     ]
    }
   ],
   "source": [
    "interaction=Array.new\n",
    "    res_intact = fetch(\"http://www.ebi.ac.uk/Tools/webservices/psicquic/intact/webservices/current/search/interactor/AT4g09010?format=tab25\");\n",
    "    agi_code = \"AT4g09010\"\n",
    "    agi_code = agi_code.upcase\n",
    "      if res_intact  # res is either the response object (Class Net::HTTPSuccess), or False, so you can test it with 'if'\n",
    "        body = res_intact.body  # get the \"body\" of the response\n",
    "        lines = body.split(\"\\n\")\n",
    "        lines.each do |int|\n",
    "          #puts \"int: #{int}\"\n",
    "          #puts puts puts\n",
    "          column = int.split(\"\\t\")\n",
    "          match_id = Regexp.new(/A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/)\n",
    "          if match_id.match(column[4])\n",
    "          idprot1 = match_id.match(column[4]).to_s.upcase\n",
    "          end\n",
    "          if match_id.match(column[5])\n",
    "          idprot2 = match_id.match(column[5]).to_s.upcase\n",
    "          end\n",
    "          if column[0]==column[1]\n",
    "            puts \"self-interaction\"\n",
    "            interaction << agi_code\n",
    "          elsif idprot1 == agi_code\n",
    "              interaction << idprot2\n",
    "          else idprot2 == agi_code\n",
    "              interaction << idprot1\n",
    "          end\n",
    "        end\n",
    "    puts interaction\n",
    "      end\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " if column[0]==column[1]\n",
    "            puts \"self-interaction\"\n",
    "            interaction << agi_code\n",
    "          elsif \n",
    "            column[4] =~ /A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/ \n",
    "            $1 == agi_code\n",
    "              column[5]  =~ /A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/\n",
    "              interaction << $1\n",
    "          else \n",
    "            column[5] =~ /A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/\n",
    "            $1 == agi_code\n",
    "              column[4] =~ /A[Tt]\\d[Gg]\\d\\d\\d\\d\\d/\n",
    "                interaction << $1\n",
    "#           end"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Ruby 2.6.3",
   "language": "ruby",
   "name": "ruby"
  },
  "language_info": {
   "file_extension": ".rb",
   "mimetype": "application/x-ruby",
   "name": "ruby",
   "version": "2.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
